Add support for HMAC type to vault_transit_secret_backend_key

roberteckert opened the pull request:
* Add key_size as it is necessary for HMAC. Default to 0 bits as that is the default for all key types _except_ HMAC currently.

<!--- See what makes a good Pull Request at : https://github.com/hashicorp/terraform-provider-vault/blob/master/.github/CONTRIBUTING.md --->


### Description
<!--- Description of the change. For example: This PR updates ABC resource so that we can XYZ --->

This PR updates the vault_transit_secret_backend_key to accept the `hmac` key type along with the requisite `key_size` argument so that transit keys of the `hmac` type can be created.

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Relates OR Closes #0000


### Checklist
- [x] Added [CHANGELOG](https://github.com/hashicorp/terraform-provider-vault/blob/master/CHANGELOG.md) entry (only for user-facing changes)
- [ ] Acceptance tests where run against all supported Vault Versions


### Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```


<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/terraform-provider-vault?pullRequest=2034)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/terraform-provider-vault?pullRequest=2034)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/terraform-provider-vault?pullRequest=2034).</sub>

fairclothjm commented on vault/resource_transit_secret_backend_key.go:
```diff
@@ -110,6 +110,20 @@ func transitSecretBackendKeyResource() *schema.Resource {
+				ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+					v := val.(int)
+					// Default for non-hmac is 0.
+					if v == 0 || v < 32 || v > 512 {
```
The transit engines allows setting a value of zero for non-hmac types. Maybe instead of checking for zero we should check that this is type `hmac`?

roberteckert replied:
I apologize for my ignorance - how does one get `type` from the validate function? I thought all you received was the key and value of the associated field. Is there a global object I can refer to here?

Alternately, I could remove this validation altogether and enforce the `hmac` rule as you propose elsewhere.

fairclothjm replied:
Woops, sorry I was thinking of DiffSuppressFunc which allows checking arbitrary `ResourceData` fields.

I think it makes sense to remove this ValidateFunc and rely on Vault's validation.

vinay-gopalan replied:
I think we can remove this validation from the TFVP. From [this bit of code](https://github.com/hashicorp/vault/blob/main/builtin/logical/transit/path_keys.go#L231-L239) in Vault I see that Vault is already doing extra validation on its end to determine if the `key_size` parameter is a valid value, and I think we can just let Vault handle dealing with the situation and returning a valid error. 

This way if you set a non-zero value in TFVP for a non-HMAC type, Vault will return an error. If you are of HMAC Key type and are inputting an invalid `key_size`, Vault will still return the right error. Suggest not worrying about extra validation of the same kind on the TFVP side, and we can remove the ValidateFunc from here :)

roberteckert replied:
Thanks for the clarification.

fairclothjm commented on vault/resource_transit_secret_backend_key.go:
```diff
@@ -221,12 +235,13 @@ func transitSecretBackendKeyCreate(d *schema.ResourceData, meta interface{}) err
 		"derived":               d.Get("derived").(bool),
 		"type":                  d.Get("type").(string),
 		"auto_rotate_period":    autoRotatePeriod,
+		"key_size":              d.Get("key_size").(int),
```
Looks like `key_size` was added in Vault 1.12. We should add a Vault Server version check in the resource and test files. You can see an examples of doing version checks in this PR: https://github.com/hashicorp/terraform-provider-vault/pull/1936/files#diff-3c100f49c020a28974d882dd5b34d276a7f825e3ccfaa18e584c1e26e2261d10R237

roberteckert replied:
I broke this out and set it separately from the block per your recommendation.

fairclothjm commented on vault/resource_transit_secret_backend_key.go:
```diff
@@ -317,24 +332,45 @@ func transitSecretBackendKeyRead(d *schema.ResourceData, meta interface{}) error
 	}
+
+	// On read, key_size will be nil if the encryption key type is not HMAC. Choosing not to set it in those cases.
+	key_size := secret.Data["key_size"]
+	if key_size != nil || secret.Data["type"] == "hmac" {
```
Same comment on checking the vault server version is 1.12 or greater which is when support was added for key_size and type hmac.

roberteckert replied:
I wrapped this whole block in a version check.

fairclothjm commented on vault/resource_transit_secret_backend_key.go:
```diff
@@ -317,24 +332,45 @@ func transitSecretBackendKeyRead(d *schema.ResourceData, meta interface{}) error
 	}
+
+	// On read, key_size will be nil if the encryption key type is not HMAC. Choosing not to set it in those cases.
+	key_size := secret.Data["key_size"]
```
nit: 
```suggestion
	keySize := secret.Data["key_size"]
```

roberteckert replied:
Fixed

fairclothjm commented on vault/resource_transit_secret_backend_key.go:
```diff
@@ -317,24 +332,45 @@ func transitSecretBackendKeyRead(d *schema.ResourceData, meta interface{}) error
+			return fmt.Errorf("expected key_size %q to be a number, and it isn't", secret.Data["key_size"])
+		}
+
+		if key_size > 0 { // Just to make sure we are not creating an HMAC type key with a key_size of 0.
```
nit: I don't think this check is necessary. We have our own ValidateFunc and Vault also will do its own validation. If for some reason our ValidateFunc fails us I think it is safe to rely on vault's validation.

roberteckert replied:
I removed this `key_size` check.

fairclothjm reviewed (commented):
Thanks! Looking good so far! Had a few comments on checking the Vault server version. 

Could we also add a [Changelog](https://github.com/hashicorp/terraform-provider-vault/blob/main/CHANGELOG.md) entry under Unreleased?

roberteckert commented:
Thanks for the feedback. I:

* Removed ValidateFunc to rely on Vault's internal validation
* Updated CHANGELOG.md
* Slightly tweaked the description of `key_size`
* Wrapped the `data["key_size"]` allocation in a check for Vault version
* Added `SkipFunc` blocks to the `hmac` tests

fairclothjm reviewed (approved):
LGTM! Thanks!

roberteckert commented on vault/resource_transit_secret_backend_key_test.go:
```diff
@@ -136,6 +136,75 @@ func TestTransitSecretBackendKey_rsa4096(t *testing.T) {
+	resourceName := "vault_transit_secret_backend_key.test"
+	resource.Test(t, resource.TestCase{
+		Providers:    testProviders,
+		PreCheck:     func() { testutil.TestAccPreCheck(t) },
```
I'll take a look now.

roberteckert commented:
This seems done - is there anything else you require from me?

fairclothjm commented:
We don't require anything else, thanks @roberteckert!
